modules/core/src/main/java/org/apache/ignite/internal/cache/query/index/IndexQueryProcessor.java

```diff
@@ -0,0 +1,449 @@
+     * a boundary, then second field isn't checked within traversing.
+     */
+    private GridCursor<IndexRow> treeIndexRange(InlineIndex idx, List<IndexQueryCriterion> criteria, int segment,
+        IndexQueryContext qryCtx) throws IgniteCheckedException {
+
+        InlineIndexRowHandler hnd = idx.segment(0).rowHandler();
+        CacheObjectContext coctx = idx.segment(0).cacheContext().cacheObjectContext();
+
+        IndexKey[] lowerBounds = new IndexKey[hnd.indexKeyDefinitions().size()];
+        IndexKey[] upperBounds = new IndexKey[hnd.indexKeyDefinitions().size()];
+
+        boolean lowerAllNulls = true;
+        boolean upperAllNulls = true;
+
+        List<RangeIndexQueryCriterion> treeCriteria = new ArrayList<>();
+
+        for (int i = 0; i < criteria.size(); i++) {
+            String f = criteria.get(i).field();
+
+            IndexKeyDefinition def = hnd.indexKeyDefinitions().get(i);
+
+            if (!def.name().equalsIgnoreCase(f))
+                throw new IgniteCheckedException("Range query doesn't match index '" + idx.name() + "'");
+
+            RangeIndexQueryCriterion c = (RangeIndexQueryCriterion) criteria.get(i);
+
+            // If index is desc, then we need to swap boundaries as user declare criteria in straight manner.
+            // For example, there is an idx (int Val desc). It means that index stores data in reverse order (1 < 0).
+            // But user won't expect for criterion gt(1) to get 0 as result, instead user will use lt(1) for getting
+            // 0. Then we need to swap user criterion.
+            if (def.order().sortOrder() == DESC)
+                c = c.swap();
+
+            treeCriteria.add(c);
+
+            IndexKey l = key(c.lower(), def, hnd.indexKeyTypeSettings(), coctx);
+            IndexKey u = key(c.upper(), def, hnd.indexKeyTypeSettings(), coctx);
+
+            if (l != null)
+                lowerAllNulls = false;
+
+            if (u != null)
+                upperAllNulls = false;
+
+            lowerBounds[i] = l;
+            upperBounds[i] = u;
+        }
+
+        IndexRow lower = lowerAllNulls ? null : new IndexSearchRowImpl(lowerBounds, hnd);
+        IndexRow upper = upperAllNulls ? null : new IndexSearchRowImpl(upperBounds, hnd);
+
+        // Step 1. Traverse index.
+        GridCursor<IndexRow> findRes = idx.find(lower, upper, segment, qryCtx);
+
+        // Step 2. Scan and filter.
+        return new GridCursor<IndexRow>() {
+            /** */
+            private final IndexRowComparator rowCmp = ((SortedIndexDefinition) idxProc.indexDefinition(idx.id())).rowComparator();
+
+            /** {@inheritDoc} */
+            @Override public boolean next() throws IgniteCheckedException {
+                if (!findRes.next())
+                    return false;
+
+                while (exclude(get(), lower, 1) || exclude(get(), upper, -1)) {
+                    if (!findRes.next())
+                        return false;
+                }
```

Comment:
This looks ambiguous.

All row in index are sorted and OR conditions are not allowed.
So, you just need to skip rows that are below a lower bound and then just detect when an upper bound is reached, is it right?

Reply:
For example we have a query `between(A, 1, 2).between(B, 3, 4)`. For this case row (A=1, B=2) is valid for our tree structure. We compare rows starting with most-significant columns. If most-significant column (A) pass comparison then less-significant columns (B) aren't checked. So B=2 is illegal by user query, but legal for our tree structure. So we need additional filter step there.

Reply:
Got it. 
Assume, there is a composite index on (A,B).
So, we need to scan contiguous range on A (top level) and filter out B (lower levels).

As I understand, idx.find(...) already returns values that satisfy condition on A, and (as you wrote) don't filter out conditions on B. 
Why we need additional checks on A? Is this because we may need to filter out rows in case of exclusive bounds?
If so, Index interface must specify whether bounds are inclusive or exclusive.

Also, the rows that returned from the Cursor are materialized, even if they will be filtered out.
It looks ok for now, but let's create a ticket for improvement.
At first glance, IndexKey might have inclusive/exclusive flag, to avoid Row materialization if range has exclusive bounds. And we can think about passing a filter.

Reply:
You pass Row (as result of get()) and Condition (search row) to exclude() method, but they have same type (IndexRow)
but comparison logic that differs from what happens inside the Index.
So, 'boundary' param name looks confusing.

Reply:
> Why we need additional checks on A

Actually don't need. I've already implement this improvement in separate branch. But I think, it's pretty simple fix, so I will do it within this PR.

> Also, the rows that returned from the Cursor are materialized

There is already a ticket for that. [IGNITE-14945](https://issues.apache.org/jira/browse/IGNITE-14945). Also I've already implement it, so just need to finish current PR to submit a new patch.

Reply:
> Is this because we may need to filter out rows in case of exclusive bounds

Yes, this is a reason. Unfortunately, currently our tree structure doesn't support exclusive bounds. It looks like a major change in tree. I think it should be done in separate ticket.

> Also, the rows that returned from the Cursor are materialized

There is already a ticket for that [IGNITE-14945](https://issues.apache.org/jira/browse/IGNITE-14945). It filters basing on inlined values. So it will speed up filtering much.